### PR TITLE
[Dynamic Shape] More correctness guards

### PR DESCRIPTION
**Pull request by Aalanli**
Add support for dynamic shape assertions in the C++ runtime.
Further, add shape and dtype check for python runtime. With this being enabled by default.

### CODE REVIEW

**Comment by Aalanli** [comment]
Moreover, there is a subtle bug where the symbol registry gets rewritten multiple times if there exists inputs which promises the same shape, but runtime inputs have different shapes.
Eg. hidet.symbol(['a', 'a']), but during runtime: hidet.randn([1, 2])

So added a check for that as well.

**Comment by Aalanli** [comment]
I have no idea why the gpt2 test passed before, there was a similar error with the llama implementation, where the compiled model reinterpreted an int64 tensor as an int32 tensor, producing random outputs.

File: python/hidet/ir/tools/printer.py [hunk 1]
```diff
@@ -474,6 +474,15 @@ def visit_Task(self, e: Task):
             Text('computations: ') + self.print_tensor_nodes(e.outputs).indent(),
             Text('attributes: {') + self({k: str(v) for k, v in e.attrs.items()}) + '}',
         ]
+        if len(e.assertions) > 0:  # between computations and attributes
+            lines.insert(
+                -1,
```
Comment (yaoyaoding) [inline python/hidet/ir/tools/printer.py:479]:
```suggestion
            lines.append(
```

File: python/hidet/option.py [hunk 2]
```diff
@@ -563,6 +570,30 @@ def debug_show_var_id(enable: bool = True):
...
+    Whether to check shapes and dtypes of all input arguments to compiled Graphs or Tasks.
+
+    Parameters
+    ----------
+    enable: bool
+        Whether to check shapes and dtypes of all input arguments to compiled Graphs or Tasks.
+    """
+    OptionContext.current().set_option('runtime_check', enable)
+
+
```
Comment (yaoyaoding) [inline python/hidet/option.py:596]: To be consistent with other option API, better to design these two as
```python
def runtime_check(enable: bool = True):
    pass
def get_runtime_check() -> bool:
    pass
```

File: python/hidet/runtime/compiled_graph.py [hunk 3]
```diff
@@ -277,6 +277,33 @@ def run_async(self, inputs):
         ret: List[hidet.Tensor]
             The output tensors.
         """
+        if hidet.option.runtime_check():
+            symbol_map = {}
+            for i, (traced, new) in enumerate(zip(self.meta.inputs, inputs)):
```
Comment (yaoyaoding) [inline python/hidet/runtime/compiled_graph.py:282]:
```suggestion
            for i, (expected, actual) in enumerate(zip(self.meta.inputs, inputs)):
```

File: python/hidet/runtime/compiled_graph.py [hunk 4]
```diff
@@ -277,6 +277,33 @@ def run_async(self, inputs):
...
+                    raise RuntimeError(
+                        f"dtype mismatch at arg {i} between original: {traced.dtype} and new: {new.dtype}"
+                    )
+                traced_shape = traced.shape
+                concrete_shape = new.shape
+                if len(traced_shape) != len(concrete_shape):
+                    raise RuntimeError(
+                        f"Rank of input {i} not equal to original. ({len(concrete_shape)} vs. {len(traced_shape)})"
+                    )
+                for j, (orig_shape, new_shape) in enumerate(zip(traced_shape, concrete_shape)):
```
Comment (yaoyaoding) [inline python/hidet/runtime/compiled_graph.py:293]:
```suggestion
                for j, (expected_size, actual_size) in enumerate(zip(traced_shape, concrete_shape)):
```
The shape usually means all dimension sizes

File: python/hidet/runtime/compiled_graph.py [hunk 5]
```diff
@@ -277,6 +277,33 @@ def run_async(self, inputs):
...
+                                           original: {orig_shape} vs. new: {new_shape}'
+                        )
+                    elif orig_shape not in symbol_map:
+                        symbol_map[orig_shape] = new_shape
+                    elif symbol_map[orig_shape] != new_shape:
+                        raise RuntimeError(
+                            f"There exists multiple instances of the same symbol {orig_shape} with\
+                            different values in inputs (ex: {symbol_map[orig_shape]} and {new_shape})"
+                        )
+
```
Comment (yaoyaoding) [inline python/hidet/runtime/compiled_graph.py:306]: Implement the checking as a new method (say `_check_inputs`).

File: python/hidet/runtime/compiled_task.py [hunk 6]
```diff
@@ -147,6 +147,35 @@ def pick_best_candidate(self, inputs, outputs) -> int:
...
+                            f'shape mismatch at dimension {j}, original: \
+                                           {orig_shape} vs. new: {new_shape}'
+                        )
+                    elif orig_shape not in symbol_map:
+                        symbol_map[orig_shape] = new_shape
+                    elif symbol_map[orig_shape] != new_shape:
+                        raise RuntimeError(
+                            f"There exists multiple instances of the same symbol {orig_shape}\
+                            with different values in inputs (ex: {symbol_map[orig_shape]} and {new_shape})"
+                        )
```
Comment (yaoyaoding) [inline python/hidet/runtime/compiled_task.py:177]: Same logic as the CompiledGraph. Consider implement as a utility function and put it at the same module of `TensorSignature`.

**Review by yaoyaoding — COMMENTED** [review]
Thanks @Aalanli.

**Comment by yaoyaoding** [comment]
Thanks @Aalanli !

**Pull request merged**